src/lab16.py

```diff
@@ -119,14 +119,20 @@ def paint_visual_effects(node, cmds, rect):
     return [transform]
 
 class ProtectedField:
-    def __init__(self, node, name, parent=None):
-        self.node = node
+    def __init__(self, obj, name, parent=None):
+        self.obj = obj
         self.name = name
         self.parent = parent
 
         self.value = None
         self.dirty = True
         self.invalidations = set()
+        self.frozen_dependencies = False
+
+    def set_dependencies(self, dependencies):
```

Comment:
I'm a little worried about the case where we call `set_dependencies` a second time on a given field. For example, suppose you have a node with `BlockLayout`, and you modify its children. Then won't we end up calling `set_dependencies` twice for the height? I agree this is something we need to do (now we depend on the new children!) but I want to make sure it'll work correctly. Have you thought carefully that all the invalidations will still be right?

Reply:
To clarify: I think it's right, but I don't 100% trust myself for this kind of code.

Reply:
The existing code has the same problem - read will keep adding things to the dependency set over and over. Is this new thing different?

Reply:
Absolutely no different. I just want to make sure we definitely have it right!

Reply:
Ok, I guess here's a question I do have: if, instead of doing `self.dependencies.add(...)`, we instead did `self.dependencies = set(...)`, that is, if instead of adding the dependencies we replaced them, would it be correct? Does it matter which one we do?

Reply:
> Ok, I guess here's a question I do have: if, instead of doing `self.dependencies.add(...)`, we instead did `self.dependencies = set(...)`, that is, if instead of adding the dependencies we replaced them, would it be correct? Does it matter which one we do?

We could do that as long as all of the dependencies were known at the same time. The current code doesn't always do that.

Reply:
FWIW I just changed the `height` dependencies to be inside the dirty bits, which at least avoids recomputing these dependencies on every layout.